File: c/epoch_bug.c
/*****************************************************************************
 *     Name: epoch_bug.c (originally ch0.time.c)
 *
 *  Summary: Epoch bug calculation.
 *
 *  Adapted: Tue 21 Aug 2001 13:46:19 (Bob Heckel--Deep C Secrets)
 *****************************************************************************
*/
#include <stdio.h>
#include <time.h>

int main(void) {
  /* 01111111 11111111 11111111 11111111 */
  time_t biggest = 0x7FFFFFFF;
  /* 10000000 00000000 00000000 00000000 */
  time_t toobig = biggest + 1;

  printf("epoch bug = %s\n", asctime(gmtime(&biggest)) );
  printf("epoch bug = %s", asctime(gmtime(&toobig)) );
}

File: scripting/epoch_bug.pl
#!/usr/bin/perl -w
##############################################################################
#    Name: epoch_bug.pl
#
# Summary: Demonstration of UNIX "2038 Bug"
#
# Adapted: Tue, 16 Jan 2001 15:53:11 (Bob Heckel -- The Best Of Dates, The
#                                     Worst Of Dates Gilbert Heaton)
##############################################################################

$EndOfTime = 0x7FFFFFFF;        # Last value valid as +signed 32-bit
$Format    = '%04d-%02d-%02d %02d:%02d:%02d' . "\n\n";

@gm = gmtime($EndOfTime);       # Break down last moment
printf("Tick, tick, tick... \t\t $Format", 
                                 $gm[5]+1900, $gm[4]+1, $gm[3], @gm[2,1,0]);

@gm = gmtime($EndOfTime + 1);   # Beyond last valid moment
printf("Kaboom! Happy New Year.\t\t  $Format", 
                                 $gm[5]+1900, $gm[4]+1, $gm[3], @gm[2,1,0]);
